src/gpio.rs

```diff
@@ -175,6 +202,89 @@ pub trait ExtiPin {
     fn check_interrupt(&mut self) -> bool;
 }
 
+impl<PIN> ExtiPin for PIN
+where
+    PIN: PinExt,
+    PIN::Mode: Interruptable,
+{
+    /// Make corresponding EXTI line sensitive to this pin
+    fn make_interrupt_source(&mut self, afio: &mut afio::Parts) {
+        let i = self.pin_id();
+        let port = self.port_id() as u32;
+        let offset = 4 * (i % 4);
+        match i {
+            0..=3 => {
+                afio.exticr1.exticr1().modify(|r, w| unsafe {
+                    w.bits((r.bits() & !(0xf << offset)) | (port << offset))
+                });
+            }
+            4..=7 => {
+                afio.exticr2.exticr2().modify(|r, w| unsafe {
+                    w.bits((r.bits() & !(0xf << offset)) | (port << offset))
+                });
+            }
+            8..=11 => {
+                afio.exticr3.exticr3().modify(|r, w| unsafe {
+                    w.bits((r.bits() & !(0xf << offset)) | (port << offset))
+                });
+            }
+            12..=15 => {
+                afio.exticr4.exticr4().modify(|r, w| unsafe {
+                    w.bits((r.bits() & !(0xf << offset)) | (port << offset))
+                });
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    /// Generate interrupt on rising edge, falling edge or both
+    fn trigger_on_edge(&mut self, exti: &EXTI, edge: Edge) {
+        let i = self.pin_id();
+        match edge {
+            Edge::RISING => {
+                exti.rtsr
+                    .modify(|r, w| unsafe { w.bits(r.bits() | (1 << i)) });
+                exti.ftsr
+                    .modify(|r, w| unsafe { w.bits(r.bits() & !(1 << i)) });
+            }
+            Edge::FALLING => {
+                exti.ftsr
+                    .modify(|r, w| unsafe { w.bits(r.bits() | (1 << i)) });
+                exti.rtsr
+                    .modify(|r, w| unsafe { w.bits(r.bits() & !(1 << i)) });
+            }
+            Edge::RISING_FALLING => {
+                exti.rtsr
+                    .modify(|r, w| unsafe { w.bits(r.bits() | (1 << i)) });
+                exti.ftsr
+                    .modify(|r, w| unsafe { w.bits(r.bits() | (1 << i)) });
+            }
```

Comment:
Looks like we now use a runtime value determine the constant to write, is that constant folded away? I.e. is the compiler smart enough to figure out that `self.pin_id` is constant?

Reply:
Yes, compiler optimize out this match to same code as before in release mode.